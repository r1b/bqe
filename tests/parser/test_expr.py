import pytest


@pytest.mark.parametrize(
    "sql",
    (
        pytest.param("SELECT column1.column2", id="dotted_column_expr_single"),
        pytest.param("SELECT column1.column2.column3", id="dotted_column_expr_multi"),
        pytest.param("SELECT column1 FROM dataset1.table1", id="dotted_table_expr_single"),
        pytest.param(
            "SELECT column1 FROM `project1`.dataset1.table1", id="dotted_table_expr_multi"
        ),
    ),
)
def test_dotted_expr_ok(sql, assert_parse_tree):
    assert_parse_tree(sql)


@pytest.mark.parametrize(
    "sql",
    (
        pytest.param("SELECT column1[0]", id="subscript_expr_integer_literal_single"),
        pytest.param("SELECT column1[0][1]", id="subscript_expr_integer_literal_multi"),
        pytest.param("SELECT column1.column2[0][1]", id="subscript_expr_path_expr_nested"),
        pytest.param("SELECT column1[OFFSET(2)]", id="subscript_expr_offset"),
        pytest.param("SELECT column1['key1']", id="subscript_expr_string_literal"),
    ),
)
def test_subscript_expr_ok(sql, assert_parse_tree):
    assert_parse_tree(sql)


@pytest.mark.parametrize(
    "sql",
    (
        pytest.param("SELECT FUNC()", id="fcall_pos_arg_none"),
        pytest.param("SELECT FUNC(1)", id="fcall_pos_arg_single"),
        pytest.param("SELECT FUNC(1, 2)", id="fcall_pos_arg_multi"),
        pytest.param("SELECT FUNC(name1 => 1)", id="fcall_named_arg_single"),
        pytest.param("SELECT FUNC(name1 => 1, name2 => 2)", id="fcall_named_arg_multi"),
        pytest.param("SELECT FUNC(1, name2 => 2)", id="fcall_kitchen_sink"),
        pytest.param("SELECT NET.IPV4_FROM_INT64(42)", id="fcall_path_expr"),
        pytest.param("SELECT FUNC(*)", id="fcall_star_expr"),
    ),
)
def test_function_call_ok(sql, assert_parse_tree):
    assert_parse_tree(sql)


@pytest.mark.parametrize(
    "sql",
    (
        pytest.param("SELECT FUNC(", id="fcall_unclosed_paren"),
        pytest.param("SELECT FUNC(,)", id="fcall_comma_only"),
        pytest.param("SELECT FUNC(,1)", id="fcall_pos_comma_leading"),
        pytest.param("SELECT FUNC(1,)", id="fcall_pos_comma_trailing"),
        pytest.param("SELECT FUNC(,name1 => 1)", id="fcall_named_comma_leading"),
        pytest.param("SELECT FUNC(name1 => 1,)", id="fcall_named_comma_trailing"),
        # TODO: Explicit error for this one
        pytest.param("SELECT FUNC(name1 => 1, 2)", id="fcall_named_before_pos"),
    ),
)
def test_function_call_error(sql, assert_parse_tree_error):
    assert_parse_tree_error(sql)


@pytest.mark.parametrize(
    "sql",
    (
        pytest.param("SELECT FUNC(DISTINCT col1)", id="agg_fcall_distinct"),
        pytest.param("SELECT FUNC(col1 IGNORE NULLS)", id="agg_fcall_ignore_nulls"),
        pytest.param("SELECT FUNC(col1 RESPECT NULLS)", id="agg_fcall_respect_nulls"),
        pytest.param("SELECT FUNC(col1 HAVING MIN col2)", id="agg_fcall_having_min"),
        pytest.param("SELECT FUNC(col1 HAVING MAX col2)", id="agg_fcall_having_max"),
        pytest.param("SELECT FUNC(col1 ORDER BY col2 DESC)", id="agg_fcall_order_by"),
        pytest.param("SELECT FUNC(col1 LIMIT 1)", id="agg_fcall_limit"),
    ),
)
def test_aggregate_function_call_ok(sql, assert_parse_tree):
    assert_parse_tree(sql)


@pytest.mark.parametrize(
    "sql",
    (
        pytest.param("SELECT x OR y", id="expr_or"),
        pytest.param("SELECT x AND y", id="expr_and"),
        pytest.param("SELECT (x OR y) OR z", id="expr_parens"),
        pytest.param("SELECT NOT x", id="expr_not"),
        pytest.param("SELECT x IS DISTINCT FROM y", id="expr_is_distinct_from"),
        pytest.param("SELECT x IS TRUE", id="expr_is"),
        pytest.param("SELECT x IS NOT NULL", id="expr_is_not"),
        pytest.param("SELECT x NOT IN UNNEST([1,2,3])", id="expr_not_in"),
        pytest.param("SELECT x IN (1,2,3)", id="expr_in"),
        pytest.param("SELECT x NOT BETWEEN y AND z", id="expr_not_between"),
        pytest.param("SELECT x BETWEEN y AND z", id="expr_between"),
        pytest.param("SELECT x NOT LIKE ALL ('a%','%b')", id="expr_not_qlike"),
        pytest.param("SELECT x LIKE ANY UNNEST(['a%','%b'])", id="expr_qlike"),
        pytest.param("SELECT x NOT LIKE 'a%'", id="expr_not_like"),
        pytest.param("SELECT x LIKE 'a%'", id="expr_like"),
        pytest.param("SELECT x != y", id="expr_ne"),
        pytest.param("SELECT x <> y", id="expr_ne_var"),
        pytest.param("SELECT x >= y", id="expr_gte"),
        pytest.param("SELECT x <= y", id="expr_lte"),
        pytest.param("SELECT x > y", id="expr_gt"),
        pytest.param("SELECT x < y", id="expr_lt"),
        pytest.param("SELECT x = y", id="expr_eq"),
        pytest.param("SELECT x | y", id="expr_bitwise_or"),
        pytest.param("SELECT x ^ y", id="expr_bitwise_xor"),
        pytest.param("SELECT x & y", id="expr_bitwise_and"),
        pytest.param("SELECT x >> y", id="expr_bitwise_shr"),
        pytest.param("SELECT x - y", id="expr_sub"),
        pytest.param("SELECT x + y", id="expr_add"),
        pytest.param("SELECT x / y", id="expr_div"),
        pytest.param("SELECT x * y", id="expr_mul"),
        pytest.param("SELECT ~x", id="expr_bitwise_not"),
        pytest.param("SELECT -x", id="expr_uminus"),
        pytest.param("SELECT +x", id="expr_uplus"),
        pytest.param("SELECT x[0]", id="expr_subscript"),
        pytest.param("SELECT x.y", id="expr_field_access"),
    ),
)
def test_column_expr_ok(sql, assert_parse_tree):
    assert_parse_tree(sql)


@pytest.mark.parametrize(
    "sql",
    (
        pytest.param("SELECT x OR y AND NOT z", id="expr_and_or_not"),
        pytest.param("SELECT (x OR y) AND NOT z", id="expr_and_or_not_parens"),
    ),
)
def test_and_or_not_ok(sql, assert_parse_tree):
    assert_parse_tree(sql)


@pytest.mark.parametrize(
    "sql", (pytest.param("SELECT x BETWEEN a AND b AND c", id="expr_between"),)
)
def test_between_ok(sql, assert_parse_tree):
    assert_parse_tree(sql)


@pytest.mark.parametrize(
    "sql",
    (
        pytest.param("SELECT ARRAY(SELECT 1)", id="keyword_subquery_array"),
        pytest.param("SELECT EXISTS(SELECT 1)", id="keyword_subquery_exists"),
    ),
)
def test_keyword_subquery_ok(sql, assert_parse_tree):
    assert_parse_tree(sql)


@pytest.mark.parametrize(
    "sql",
    (
        pytest.param("SELECT CASE WHEN 1 THEN 1 END", id="case_single"),
        pytest.param("SELECT CASE WHEN 1 THEN 1 WHEN 2 THEN 2 END", id="case_multi"),
        pytest.param("SELECT CASE WHEN 1 THEN 1 WHEN 2 THEN 2 ELSE 3 END", id="case_multi_else"),
        pytest.param("SELECT CASE col1 WHEN 1 THEN 1 END", id="case_match_single"),
        pytest.param("SELECT CASE col1 WHEN 1 THEN 1 WHEN 2 THEN 2 END", id="case_match_multi"),
        pytest.param(
            "SELECT CASE col1 WHEN 1 THEN 1 WHEN 2 THEN 2 ELSE 3 END", id="case_match_multi_else"
        ),
    ),
)
def test_case_ok(sql, assert_parse_tree):
    assert_parse_tree(sql)
