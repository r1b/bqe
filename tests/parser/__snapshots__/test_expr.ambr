# serializer version: 1
# name: test_aggregate_function_call_ok[agg_fcall_distinct]
  '''
  select
    select_list
      select_expr
        function_call_expr
          function_name
            ident	FUNC
          function_distinct
          function_args
            pos_arg
              ident	col1
  
  '''
# ---
# name: test_aggregate_function_call_ok[agg_fcall_having_max]
  '''
  select
    select_list
      select_expr
        function_call_expr
          function_name
            ident	FUNC
          function_args
            pos_arg
              ident	col1
          function_modifiers
            having_modifier
              having_max
              ident	col2
  
  '''
# ---
# name: test_aggregate_function_call_ok[agg_fcall_having_min]
  '''
  select
    select_list
      select_expr
        function_call_expr
          function_name
            ident	FUNC
          function_args
            pos_arg
              ident	col1
          function_modifiers
            having_modifier
              having_min
              ident	col2
  
  '''
# ---
# name: test_aggregate_function_call_ok[agg_fcall_ignore_nulls]
  '''
  select
    select_list
      select_expr
        function_call_expr
          function_name
            ident	FUNC
          function_args
            pos_arg
              ident	col1
          function_modifiers
            nulls_modifier
              ignore_nulls
  
  '''
# ---
# name: test_aggregate_function_call_ok[agg_fcall_limit]
  '''
  select
    select_list
      select_expr
        function_call_expr
          function_name
            ident	FUNC
          function_args
            pos_arg
              ident	col1
          function_modifiers
            limit_modifier
              integer_literal	1
  
  '''
# ---
# name: test_aggregate_function_call_ok[agg_fcall_order_by]
  '''
  select
    select_list
      select_expr
        function_call_expr
          function_name
            ident	FUNC
          function_args
            pos_arg
              ident	col1
          function_modifiers
            order_by_modifier
              order_by_items
                order_by_item
                  ident	col2
                  order_by_item_sort	DESC
  
  '''
# ---
# name: test_aggregate_function_call_ok[agg_fcall_respect_nulls]
  '''
  select
    select_list
      select_expr
        function_call_expr
          function_name
            ident	FUNC
          function_args
            pos_arg
              ident	col1
          function_modifiers
            nulls_modifier
              respect_nulls
  
  '''
# ---
# name: test_and_or_not_ok[expr_and_or_not]
  '''
  select
    select_list
      select_expr
        or_expr
          ident	x
          and_expr
            ident	y
            not_expr
              ident	z
  
  '''
# ---
# name: test_and_or_not_ok[expr_and_or_not_parens]
  '''
  select
    select_list
      select_expr
        and_expr
          parenthesized_expr
            or_expr
              ident	x
              ident	y
          not_expr
            ident	z
  
  '''
# ---
# name: test_between_ok[expr_between]
  '''
  select
    select_list
      select_expr
        and_expr
          between_expr
            ident	x
            ident	a
            ident	b
          ident	c
  
  '''
# ---
# name: test_case_ok[case_match_multi]
  '''
  select
    select_list
      select_expr
        case_match_expr
          ident	col1
          when_then_expr
            integer_literal	1
            integer_literal	1
          when_then_expr
            integer_literal	2
            integer_literal	2
  
  '''
# ---
# name: test_case_ok[case_match_multi_else]
  '''
  select
    select_list
      select_expr
        case_match_expr
          ident	col1
          when_then_expr
            integer_literal	1
            integer_literal	1
          when_then_expr
            integer_literal	2
            integer_literal	2
          else_expr
            integer_literal	3
  
  '''
# ---
# name: test_case_ok[case_match_single]
  '''
  select
    select_list
      select_expr
        case_match_expr
          ident	col1
          when_then_expr
            integer_literal	1
            integer_literal	1
  
  '''
# ---
# name: test_case_ok[case_multi]
  '''
  select
    select_list
      select_expr
        case_expr
          when_then_expr
            integer_literal	1
            integer_literal	1
          when_then_expr
            integer_literal	2
            integer_literal	2
  
  '''
# ---
# name: test_case_ok[case_multi_else]
  '''
  select
    select_list
      select_expr
        case_expr
          when_then_expr
            integer_literal	1
            integer_literal	1
          when_then_expr
            integer_literal	2
            integer_literal	2
          else_expr
            integer_literal	3
  
  '''
# ---
# name: test_case_ok[case_single]
  '''
  select
    select_list
      select_expr
        case_expr
          when_then_expr
            integer_literal	1
            integer_literal	1
  
  '''
# ---
# name: test_column_expr_ok[expr_add]
  '''
  select
    select_list
      select_expr
        add_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_and]
  '''
  select
    select_list
      select_expr
        and_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_between]
  '''
  select
    select_list
      select_expr
        between_expr
          ident	x
          ident	y
          ident	z
  
  '''
# ---
# name: test_column_expr_ok[expr_bitwise_and]
  '''
  select
    select_list
      select_expr
        bitwise_and_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_bitwise_not]
  '''
  select
    select_list
      select_expr
        bitwise_not_expr
          ident	x
  
  '''
# ---
# name: test_column_expr_ok[expr_bitwise_or]
  '''
  select
    select_list
      select_expr
        bitwise_or_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_bitwise_shr]
  '''
  select
    select_list
      select_expr
        bitwise_shr_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_bitwise_xor]
  '''
  select
    select_list
      select_expr
        bitwise_xor_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_div]
  '''
  select
    select_list
      select_expr
        div_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_eq]
  '''
  select
    select_list
      select_expr
        eq_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_field_access]
  '''
  select
    select_list
      select_expr
        ident
          x
          y
  
  '''
# ---
# name: test_column_expr_ok[expr_gt]
  '''
  select
    select_list
      select_expr
        gt_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_gte]
  '''
  select
    select_list
      select_expr
        gte_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_in]
  '''
  select
    select_list
      select_expr
        in_expr
          ident	x
          parenthesized_list
            integer_literal	1
            integer_literal	2
            integer_literal	3
  
  '''
# ---
# name: test_column_expr_ok[expr_is]
  '''
  select
    select_list
      select_expr
        is_expr
          ident	x
          boolean_literal	TRUE
  
  '''
# ---
# name: test_column_expr_ok[expr_is_distinct_from]
  '''
  select
    select_list
      select_expr
        is_distinct_from_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_is_not]
  '''
  select
    select_list
      select_expr
        is_not_expr
          ident	x
          null_literal	NULL
  
  '''
# ---
# name: test_column_expr_ok[expr_like]
  '''
  select
    select_list
      select_expr
        like_expr
          ident	x
          string_literal	'a%'
  
  '''
# ---
# name: test_column_expr_ok[expr_lt]
  '''
  select
    select_list
      select_expr
        lt_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_lte]
  '''
  select
    select_list
      select_expr
        lte_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_mul]
  '''
  select
    select_list
      select_expr
        mul_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_ne]
  '''
  select
    select_list
      select_expr
        ne_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_ne_var]
  '''
  select
    select_list
      select_expr
        ne_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_not]
  '''
  select
    select_list
      select_expr
        not_expr
          ident	x
  
  '''
# ---
# name: test_column_expr_ok[expr_not_between]
  '''
  select
    select_list
      select_expr
        not_between_expr
          ident	x
          ident	y
          ident	z
  
  '''
# ---
# name: test_column_expr_ok[expr_not_in]
  '''
  select
    select_list
      select_expr
        not_in_expr
          ident	x
          unnest_expr
            array_literal
              integer_literal	1
              integer_literal	2
              integer_literal	3
  
  '''
# ---
# name: test_column_expr_ok[expr_not_like]
  '''
  select
    select_list
      select_expr
        not_like_expr
          ident	x
          string_literal	'a%'
  
  '''
# ---
# name: test_column_expr_ok[expr_not_qlike]
  '''
  select
    select_list
      select_expr
        not_qlike_expr
          ident	x
          qlike_quantifier	ALL
          parenthesized_list
            string_literal	'a%'
            string_literal	'%b'
  
  '''
# ---
# name: test_column_expr_ok[expr_or]
  '''
  select
    select_list
      select_expr
        or_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_parens]
  '''
  select
    select_list
      select_expr
        or_expr
          parenthesized_expr
            or_expr
              ident	x
              ident	y
          ident	z
  
  '''
# ---
# name: test_column_expr_ok[expr_qlike]
  '''
  select
    select_list
      select_expr
        qlike_expr
          ident	x
          qlike_quantifier	ANY
          unnest_expr
            array_literal
              string_literal	'a%'
              string_literal	'%b'
  
  '''
# ---
# name: test_column_expr_ok[expr_sub]
  '''
  select
    select_list
      select_expr
        sub_expr
          ident	x
          ident	y
  
  '''
# ---
# name: test_column_expr_ok[expr_subscript]
  '''
  select
    select_list
      select_expr
        subscript_expr
          ident	x
          integer_literal	0
  
  '''
# ---
# name: test_column_expr_ok[expr_uminus]
  '''
  select
    select_list
      select_expr
        uminus_expr
          ident	x
  
  '''
# ---
# name: test_column_expr_ok[expr_uplus]
  '''
  select
    select_list
      select_expr
        uplus_expr
          ident	x
  
  '''
# ---
# name: test_dotted_expr_ok[dotted_column_expr_multi]
  '''
  select
    select_list
      select_expr
        ident
          column1
          column2
          column3
  
  '''
# ---
# name: test_dotted_expr_ok[dotted_column_expr_single]
  '''
  select
    select_list
      select_expr
        ident
          column1
          column2
  
  '''
# ---
# name: test_dotted_expr_ok[dotted_table_expr_multi]
  '''
  select
    select_list
      select_expr
        ident	column1
    from_clause
      from_item_table
        ident
          `project1`
          dataset1
          table1
  
  '''
# ---
# name: test_dotted_expr_ok[dotted_table_expr_single]
  '''
  select
    select_list
      select_expr
        ident	column1
    from_clause
      from_item_table
        ident
          dataset1
          table1
  
  '''
# ---
# name: test_function_call_ok[fcall_kitchen_sink]
  '''
  select
    select_list
      select_expr
        function_call_expr
          function_name
            ident	FUNC
          function_args
            pos_arg
              integer_literal	1
            named_arg
              ident	name2
              integer_literal	2
  
  '''
# ---
# name: test_function_call_ok[fcall_named_arg_multi]
  '''
  select
    select_list
      select_expr
        function_call_expr
          function_name
            ident	FUNC
          function_args
            named_arg
              ident	name1
              integer_literal	1
            named_arg
              ident	name2
              integer_literal	2
  
  '''
# ---
# name: test_function_call_ok[fcall_named_arg_single]
  '''
  select
    select_list
      select_expr
        function_call_expr
          function_name
            ident	FUNC
          function_args
            named_arg
              ident	name1
              integer_literal	1
  
  '''
# ---
# name: test_function_call_ok[fcall_path_expr]
  '''
  select
    select_list
      select_expr
        function_call_expr
          function_name
            ident
              NET
              IPV4_FROM_INT64
          function_args
            pos_arg
              integer_literal	42
  
  '''
# ---
# name: test_function_call_ok[fcall_pos_arg_multi]
  '''
  select
    select_list
      select_expr
        function_call_expr
          function_name
            ident	FUNC
          function_args
            pos_arg
              integer_literal	1
            pos_arg
              integer_literal	2
  
  '''
# ---
# name: test_function_call_ok[fcall_pos_arg_none]
  '''
  select
    select_list
      select_expr
        function_call_expr
          function_name
            ident	FUNC
  
  '''
# ---
# name: test_function_call_ok[fcall_pos_arg_single]
  '''
  select
    select_list
      select_expr
        function_call_expr
          function_name
            ident	FUNC
          function_args
            pos_arg
              integer_literal	1
  
  '''
# ---
# name: test_function_call_ok[fcall_star_expr]
  '''
  select
    select_list
      select_expr
        function_call_expr
          function_name
            ident	FUNC
          function_args
            star_arg
  
  '''
# ---
# name: test_keyword_subquery_ok[keyword_subquery_array]
  '''
  select
    select_list
      select_expr
        array_subquery_expr
          select
            select_list
              select_expr
                integer_literal	1
  
  '''
# ---
# name: test_keyword_subquery_ok[keyword_subquery_exists]
  '''
  select
    select_list
      select_expr
        exists_subquery_expr
          select
            select_list
              select_expr
                integer_literal	1
  
  '''
# ---
# name: test_subscript_expr_ok[subscript_expr_integer_literal_multi]
  '''
  select
    select_list
      select_expr
        subscript_expr
          subscript_expr
            ident	column1
            integer_literal	0
          integer_literal	1
  
  '''
# ---
# name: test_subscript_expr_ok[subscript_expr_integer_literal_single]
  '''
  select
    select_list
      select_expr
        subscript_expr
          ident	column1
          integer_literal	0
  
  '''
# ---
# name: test_subscript_expr_ok[subscript_expr_offset]
  '''
  select
    select_list
      select_expr
        subscript_expr
          ident	column1
          function_call_expr
            function_name
              ident	OFFSET
            function_args
              pos_arg
                integer_literal	2
  
  '''
# ---
# name: test_subscript_expr_ok[subscript_expr_path_expr_nested]
  '''
  select
    select_list
      select_expr
        subscript_expr
          subscript_expr
            ident
              column1
              column2
            integer_literal	0
          integer_literal	1
  
  '''
# ---
# name: test_subscript_expr_ok[subscript_expr_string_literal]
  '''
  select
    select_list
      select_expr
        subscript_expr
          ident	column1
          string_literal	'key1'
  
  '''
# ---
